### PR TITLE
atomic ordering docs

### PR DESCRIPTION
**Pull request by RalfJung**
Discussion in https://github.com/rust-lang/rfcs/pull/2503 revealed that this could be improved. I hope this helps.

### CODE REVIEW

File: src/libcore/sync/atomic.rs [hunk 1]
```diff
@@ -384,9 +419,11 @@ impl AtomicBool {
     /// was updated.
     ///
     /// `compare_and_swap` also takes an [`Ordering`] argument which describes the memory
-    /// ordering of this operation.
+    /// ordering of this operation. Notice that even when using [`AcqRel`], the operation
+    /// might fail and hence just perform an `Acquire` load, but not have `Release` semantics.
```
Comment (stjepang) [inline src/libcore/sync/atomic.rs]: This also applies to `compare_exchange` and `compare_exchange_weak`.

**Review by ghost — COMMENTED** [review]
I like these clarifications! 👍

Another thing we could do is specify which orderings are valid for each operation (in the docs of each of the methods like `store`, `load`, `swap`, etc.).

File: src/libcore/sync/atomic.rs (hunk 1 above)
Comment (RalfJung) [reply src/libcore/sync/atomic.rs]: No it does not. Those have a separate argument for the ordering used in case of failure.

**Comment by RalfJung** [comment]
> Another thing we could do is specify which orderings are valid for each operation (in the docs of each of the methods like store, load, swap, etc.).

Hm... that would make for quite a large diff, and it is already stated in the `Ordering` enum. However, I also considered saying for each RMW operation that `Release` and `Acquire` actually perform a `Relaxed` access here. So I guess you are right.

File: src/libcore/sync/atomic.rs [hunk 2]
```diff
@@ -193,32 +193,67 @@ pub enum Ordering {
...
     ///
+    /// Notice that using this ordering for an operation that combines loads
+    /// and stores leads to a [`Relaxed`] load operation!
+    ///
+    /// This ordering is only applicable for operations that can perform a store.
+    ///
+    /// Corresponds to LLVM's [`Release`] ordering.
+    ///
+    /// [`Release`]: http://llvm.org/docs/Atomics.html#release
     /// [`Acquire`]: http://llvm.org/docs/Atomics.html#acquire
```
Comment (kennytm) [inline src/libcore/sync/atomic.rs]: Could you change these links to HTTPS as well?

File: src/libcore/sync/atomic.rs [hunk 3]
```diff
@@ -384,9 +419,11 @@ impl AtomicBool {
     /// was updated.
     ///
     /// `compare_and_swap` also takes an [`Ordering`] argument which describes the memory
-    /// ordering of this operation.
+    /// ordering of this operation. Notice that even when using [`AcqRel`], the operation
+    /// might fail and hence just perform an `Acquire` load, but not have `Release` semantics.
     ///
     /// [`Ordering`]: enum.Ordering.html
+    /// [`Ordering`]: enum.Ordering.html#variant.AcqRel
```
Comment (kennytm) [inline src/libcore/sync/atomic.rs]: The label should be ``[`AcqRel`]``

**Comment by RalfJung** [comment]
> Another thing we could do is specify which orderings are valid for each operation (in the docs of each of the methods like store, load, swap, etc.).

Done. I hope I didn't miss any, nor screw up one of them.^^

**Comment by TimNN** [comment]
@stjepang: Since you already commented, and are much more familiar with the subject the I am, I think, could you review this PR?

**Review by ghost — APPROVED** [review]
I've read through the whole set of changes and everything seems correct and clear. 👍

**Pull request merged**